SqlClient managed networking improvements

Wraith2 opened the pull request:
While profiling SqlClient for various other PR's it became clear that the managed implementation was wasting a lot of resources in networking and that this was severely degrading performance relative to the native pinvoke version. This PR works towards closing that gap.

The core of the changes start in SNIPacket. This has had a start field added which allows the ability to reserve space at the start of the packet for an additional header, the MARS smux function can now take advantage of this reservation removing the need to allocate a new packet and data array before sending. As a result of the header functionality it is no longer possible to cache and re-use a single attention packet.

The lifetime management of packets has been tightened and calls to `Release` added to avoid dropping packets and their rented buffers for the GC to deal with. The packet has had the IDispose and IEquatable removed because they are not needed and the explicit call to Release over Dispose makes the lifetime management much clearer in the code. As a result of the clearer packet management there is now no need to clone a packet before sending which avoids further re-allocation. This changes the number of packet copies from 2 to 1 in standard cases and with the smux reservation 3 to 1 for mars connections.

The functional and manual tests run through sucessfully.

Benchmark results are overall positive:

|           Method |     Mean |     Error |    StdDev | Gen 0/1k Op | Gen 1/1k Op | Gen 2/1k Op | Allocated Memory/Op |
|----------------- |---------:|----------:|----------:|------------:|------------:|------------:|--------------------:|
| master native    | 75.37 ms | 0.5611 ms | 0.5249 ms |           - |           - |           - |           273.97 KB |
| master managed   | 77.81 ms | 0.7198 ms | 0.6733 ms |   7000.0000 |           - |           - |           318.35 KB |
| branch managed   | 74.07 ms | 0.2619 ms | 0.2322 ms |           - |           - |           - |           298.19 KB |

Those results make the managed version looks very similat to the native one in performance, this isn't really the case. Profiles show the improvement in a slighlty more nuanced way. 

Baseline, master native: very little GC activity and the test takes 10s
![master native](https://user-images.githubusercontent.com/13322696/52888147-7c1fc380-3172-11e9-8aec-c970505cda1f.PNG)

Current, master managed: huge amounts of gc activity, a gen0 roughly every 16ms, the test takes 13s
![master managed](https://user-images.githubusercontent.com/13322696/52888154-817d0e00-3172-11e9-9222-c422c767354b.PNG)

Improved, branch managed: much less GC activity and the test takes 12s
![smux managed](https://user-images.githubusercontent.com/13322696/52888158-8641c200-3172-11e9-81dc-3499e75bedb8.PNG)

So looking at it optimistically the memory improvements pull back roughly a third of the speed difference between managed and native.

/cc @afsanehr, @tarikulsabbir, @david-engel, @saurabh500

Wraith2 committed:
rework SNIPacket and usage

Wraith2 committed:
Merge remote-tracking branch 'dotnet/master' into sqlperf-managedsmux

karelz commented:
@AfsanehR @tarikulsabbir @Gary-Zh @David-Engel can you please look at this one too? Also 2 weeks old without code review :(

AfsanehR commented:
@karelz  @Wraith2  We are working towards an internal deadline and we will start reviewing these PRs starting next week.

Wraith2 committed:
Merge remote-tracking branch 'dotnet/master' into sqlperf-managedsmux

karelz commented:
@AfsanehR any update on review ETA?

Gary-Zh commented:
@karelz  We are reviewing this PR.

Wraith2 committed:
Merge remote-tracking branch 'dotnet/master' into sqlperf-managedsmux

Wraith2 committed:
Merge remote-tracking branch 'dotnet/master' into sqlperf-managedsmux

Gary-Zh commented on src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNIPacket.cs:
```diff
@@ -259,48 +238,10 @@ public void ReadFromStream(Stream stream)
-        /// </summary>
-        /// <param name="obj"></param>
-        /// <returns>true if equal</returns>
-        public override bool Equals(object obj)
```
What's the purpose to remove these public functions?

Wraith2 replied:
Packets are never compared to each other. If in further work they are compared it will be in a pool and the reference equality will be all that is needed. As such these methods aren't needed or used currently.

Gary-Zh replied:
I suggest to remain what we have before even if it's not used currently.
There should be a reason why they were added.

Wraith2 replied:
They were originally added with the intention of using them when pooling managed packets. The packet pooling was never fully written and integrated so they weren't used. See the removed `WritePacketCache` code in `TdsParserStateObjectManaged`. Are you sure you want dead and trivial code keeping?

Gary-Zh commented on src/System.Data.SqlClient/src/System/Data/SqlClient/TdsParserStateObjectFactory.Windows.cs:
```diff
@@ -19,14 +19,14 @@ internal sealed class TdsParserStateObjectFactory
         //public static bool UseManagedSNI { get; } = AppContext.TryGetSwitch(UseLegacyNetworkingOnWindows, out shouldUseLegacyNetorking) ? !shouldUseLegacyNetorking : true;
 
-#if DEBUG
+//#if DEBUG
```
Is this to take the managed SNIPackat on windows? For test purpose?

Wraith2 replied:
Yes, it allows me to switch modes for testing and profiling without having to manually edit the nuget file which is slow and tedious when you're trying to iterate on a problem. Since it's only active in debug builds which will never be in the published packages it shouldn't cause a problem.

saurabh500 replied:
Wouldn't this be available in release builds now that you have commented out the ifdef?

Wraith2 replied:
Yes. At some point it would be useful to re-enable the ability to switch modes on windows but I don't know when that point will be. I can reset this to be debug only again if you prefer.

saurabh500 replied:
Yeah this should be debug only. I don't want this to be exposed in Release build unless we want to allow the product being shipped, to be able to do the switch with env variables. 
I definitely understand why you need this switch for your testing of release build and that is reasonable. But please exclude the changes from getting in Release build.

Wraith2 replied:
Ok, done.

Gary-Zh commented on src/System.Data.SqlClient/src/System/Data/SqlClient/TdsParserStateObjectManaged.cs:
```diff
@@ -162,13 +144,17 @@ internal override PacketHandle ReadAsync(SessionHandle handle, out uint error)
+            //    SetPacketData(PacketHandle.FromManagedPacket(attnPacket), SQL.AttentionHeader, TdsEnums.HEADER_LEN);
+            //    _sniAsyncAttnPacket = attnPacket;
+            //}
+            //return PacketHandle.FromManagedPacket(_sniAsyncAttnPacket);
```
commented out code.

Wraith2 replied:
Removed.

Wraith2 commented on src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNIHandle.cs:
```diff
@@ -84,6 +84,8 @@ internal abstract class SNIHandle
         /// </summary>
         public abstract Guid ConnectionId { get; }
 
+        public virtual bool SMUXEnabled => false;
```
It isn't publc, it's a public method on an internal class so it's effectively private to external view. it is however part of the public surface inside the assembly.

Wraith2 committed:
address feedback

Wraith2 commented on src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNIPacket.cs:
```diff
@@ -12,37 +13,28 @@ namespace System.Data.SqlClient.SNI
-                _description = value;
-            }
+        public SNIPacket(int capacity, bool reserveMuxHeader=false)
+		{
```
Fixed.

Wraith2 commented on src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNIPacket.cs:
```diff
@@ -58,15 +50,11 @@ public string Description
         /// Packet validity
         /// </summary>
-        public bool IsInvalid => (_data == null);
+        public bool IsInvalid => _data is null;
```
Same as above, the final accessibility will be private but i use public because even internally the distinction between public-internal api surface and private implementation is important to me. Using internal everywhre means you always lose encapsulation.

Gary-Zh replied:
I see

Wraith2 commented on src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNIPacket.cs:
```diff
@@ -211,36 +175,51 @@ public int TakeData(byte[] buffer, int dataOffset, int size)
             }
 
-            Buffer.BlockCopy(_data, _offset, buffer, dataOffset, size);
+            Buffer.BlockCopy(_data, (_header + _offset), buffer, dataOffset, size);
```
Removed.

Wraith2 committed:
Merge remote-tracking branch 'dotnet/master' into sqlperf-managedsmux

Wraith2 committed:
remove debug ifdef

Gary-Zh commented:
@Wraith2  I've run EFCore test on Windows and some of the test cases failed due to server timeout.
Here's the stack trace: 
```
System.Data.SqlClient.SqlException : Timeout expired.  The timeout period elapsed prior to completion of the operation or the server is not responding.\r\n---- System.ComponentModel.Win32Exception : The wait operation timed out.
   at System.Data.SqlClient.SqlConnection.OnError(SqlException exception, Boolean breakConnection, Action`1 wrapCloseInAction)
   at System.Data.SqlClient.SqlInternalConnection.OnError(SqlException exception, Boolean breakConnection, Action`1 wrapCloseInAction)
   at System.Data.SqlClient.TdsParser.ThrowExceptionAndWarning(TdsParserStateObject stateObj, Boolean callerHasConnectionLock, Boolean asyncClose)
   at System.Data.SqlClient.TdsParser.TryRun(RunBehavior runBehavior, SqlCommand cmdHandler, SqlDataReader dataStream, BulkCopySimpleResultSet bulkCopyHandler, TdsParserStateObject stateObj, Boolean& dataReady)
   at System.Data.SqlClient.SqlDataReader.TryConsumeMetaData()
   at System.Data.SqlClient.SqlDataReader.get_MetaData()
   at System.Data.SqlClient.SqlCommand.FinishExecuteReader(SqlDataReader ds, RunBehavior runBehavior, String resetOptionsString)
   at System.Data.SqlClient.SqlCommand.RunExecuteReaderTds(CommandBehavior cmdBehavior, RunBehavior runBehavior, Boolean returnStream, Boolean async, Int32 timeout, Task& task, Boolean asyncWrite, SqlDataReader ds)
   at System.Data.SqlClient.SqlCommand.RunExecuteReader(CommandBehavior cmdBehavior, RunBehavior runBehavior, Boolean returnStream, String method)
   at System.Data.SqlClient.SqlCommand.ExecuteScalar()
   at Microsoft.EntityFrameworkCore.SqlServer.Scaffolding.Internal.SqlServerDatabaseModelFactory.GetDefaultSchema(DbConnection connection) in D:\dotnetcore\EFCoreTest\src\EFCore.SqlServer\Scaffolding\Internal\SqlServerDatabaseModelFactory.cs:line 183
   at Microsoft.EntityFrameworkCore.SqlServer.Scaffolding.Internal.SqlServerDatabaseModelFactory.Create(DbConnection connection, DatabaseModelFactoryOptions options) in D:\dotnetcore\EFCoreTest\src\EFCore.SqlServer\Scaffolding\Internal\SqlServerDatabaseModelFactory.cs:line 121
   at Microsoft.EntityFrameworkCore.TestUtilities.RelationalDatabaseCleaner.Clean(DatabaseFacade facade) in D:\dotnetcore\EFCoreTest\test\EFCore.Relational.Specification.Tests\TestUtilities\RelationalDatabaseCleaner.cs:line 52
   at Microsoft.EntityFrameworkCore.TestUtilities.SqlServerDatabaseFacadeExtensions.<>c.<EnsureClean>b__0_0(DatabaseFacade database) in D:\dotnetcore\EFCoreTest\test\EFCore.SqlServer.FunctionalTests\TestUtilities\SqlServerDatabaseFacadeExtensions.cs:line 12
   at Microsoft.EntityFrameworkCore.ExecutionStrategyExtensions.<>c__2`1.<Execute>b__2_0(<>f__AnonymousType0`2 s) in D:\dotnetcore\EFCoreTest\src\EFCore\Extensions\ExecutionStrategyExtensions.cs:line 77
   at Microsoft.EntityFrameworkCore.ExecutionStrategyExtensions.<>c__DisplayClass12_0`2.<Execute>b__0(DbContext c, TState s) in D:\dotnetcore\EFCoreTest\src\EFCore\Extensions\ExecutionStrategyExtensions.cs:line 354
   at Microsoft.EntityFrameworkCore.SqlServer.Storage.Internal.SqlServerExecutionStrategy.Execute[TState,TResult](TState state, Func`3 operation, Func`3 verifySucceeded) in D:\dotnetcore\EFCoreTest\src\EFCore.SqlServer\Storage\Internal\SqlServerExecutionStrategy.cs:line 47
   at Microsoft.EntityFrameworkCore.ExecutionStrategyExtensions.Execute[TState,TResult](IExecutionStrategy strategy, Func`2 operation, Func`2 verifySucceeded, TState state) in D:\dotnetcore\EFCoreTest\src\EFCore\Extensions\ExecutionStrategyExtensions.cs:line 352
   at Microsoft.EntityFrameworkCore.ExecutionStrategyExtensions.Execute[TState,TResult](IExecutionStrategy strategy, TState state, Func`2 operation) in D:\dotnetcore\EFCoreTest\src\EFCore\Extensions\ExecutionStrategyExtensions.cs:line 302
   at Microsoft.EntityFrameworkCore.ExecutionStrategyExtensions.Execute[TState](IExecutionStrategy strategy, TState state, Action`1 operation) in D:\dotnetcore\EFCoreTest\src\EFCore\Extensions\ExecutionStrategyExtensions.cs:line 70
   at Microsoft.EntityFrameworkCore.TestUtilities.SqlServerDatabaseFacadeExtensions.EnsureClean(DatabaseFacade databaseFacade) in D:\dotnetcore\EFCoreTest\test\EFCore.SqlServer.FunctionalTests\TestUtilities\SqlServerDatabaseFacadeExtensions.cs:line 11
   at Microsoft.EntityFrameworkCore.TestUtilities.SqlServerTestStore.Clean(DbContext context) in D:\dotnetcore\EFCoreTest\test\EFCore.SqlServer.FunctionalTests\TestUtilities\SqlServerTestStore.cs:line 136
   at Microsoft.EntityFrameworkCore.TestUtilities.SqlServerTestStore.CreateDatabase() in D:\dotnetcore\EFCoreTest\test\EFCore.SqlServer.FunctionalTests\TestUtilities\SqlServerTestStore.cs:line 119
   at Microsoft.EntityFrameworkCore.TestUtilities.SqlServerTestStore.Initialize(Func`1 createContext, Action`1 seed) in D:\dotnetcore\EFCoreTest\test\EFCore.SqlServer.FunctionalTests\TestUtilities\SqlServerTestStore.cs:line 80
   at Microsoft.EntityFrameworkCore.TestUtilities.TestStore.Initialize(IServiceProvider serviceProvider, Func`1 createContext, Action`1 seed) in D:\dotnetcore\EFCoreTest\test\EFCore.Specification.Tests\TestUtilities\TestStore.cs:line 44
   at Microsoft.EntityFrameworkCore.TestUtilities.RelationalTestStore.Initialize(IServiceProvider serviceProvider, Func`1 createContext, Action`1 seed) in D:\dotnetcore\EFCoreTest\test\EFCore.Relational.Specification.Tests\TestUtilities\RelationalTestStore.cs:line 29
   at Microsoft.EntityFrameworkCore.TestUtilities.SqlServerTestStore.InitializeSqlServer(IServiceProvider serviceProvider, Func`1 createContext, Action`1 seed) in D:\dotnetcore\EFCoreTest\test\EFCore.SqlServer.FunctionalTests\TestUtilities\SqlServerTestStore.cs:line 72
   at Microsoft.EntityFrameworkCore.TestUtilities.SqlServerTestStore.CreateInitialized(String name, Boolean useFileName, Nullable`1 multipleActiveResultSets) in D:\dotnetcore\EFCoreTest\test\EFCore.SqlServer.FunctionalTests\TestUtilities\SqlServerTestStore.cs:line 42
   at Microsoft.EntityFrameworkCore.Query.QueryBugsTest.CreateTestStore[TContext](Func`1 contextCreator, Action`1 contextInitializer) in D:\dotnetcore\EFCoreTest\test\EFCore.SqlServer.FunctionalTests\Query\QueryBugsTest.cs:line 5358
   at Microsoft.EntityFrameworkCore.Query.QueryBugsTest.CreateDatabase9277() in D:\dotnetcore\EFCoreTest\test\EFCore.SqlServer.FunctionalTests\Query\QueryBugsTest.cs:line 2737
   at Microsoft.EntityFrameworkCore.Query.QueryBugsTest.From_sql_gets_value_of_out_parameter_in_stored_procedure() in D:\dotnetcore\EFCoreTest\test\EFCore.SqlServer.FunctionalTests\Query\QueryBugsTest.cs:line 2705
----- Inner Stack Trace -----
```
Failed test cases are: 
`Microsoft.EntityFrameworkCore.Query.QueryBugsTest.From_sql_gets_value_of_out_parameter_in_stored_procedure`
and 
`Microsoft.EntityFrameworkCore.SequenceEndToEndTest.Can_use_sequence_end_to_end_from_multiple_contexts_concurrently_async`

Wraith2 commented:
Interesting. I saw similar behaviour when I enabled packet pooling but couldn't work out why. Nothing I've changed should affect the flow of packets only how they're constructed, why would a timeout occur? Can you see anything I've changed anywhere that could cause this?

saurabh500 commented on src/System.Data.SqlClient/src/System/Data/SqlClient/TdsParserStateObjectManaged.cs:
```diff
@@ -81,39 +77,25 @@ protected override void RemovePacketFromPendingList(PacketHandle packet)
-            if (null != sessionHandle || null != packetHandle)
+            if (sessionHandle != null)
             {
-                packetHandle?.Dispose();
```
Where is the packetHandle being disposed in the new code?

saurabh500 replied:
NVM, I didnt realize I am looking at TdsParserStateObjectManaged where you have changed the lifecycle of packet.

Wraith2 replied:
I've tried to make sure that every packet is correctly released which means there are now Release calls scattered throughout the code. It isn't the end of the world is a packet is dropped without clearing but it'd rather it didn't happen.

As I mentioned my eventual intention is to pool the packets so that a small number are kept and reused for all connections as this will substantially reduce allocations.

saurabh500 replied:
Yeah I understand. I didn't clarify. It seemed to me that I was read the TPSONative code and it seemed that the dispose logic which existed earlier, has been removed. Later I realized that this is managed TDSO where you are taking care of the cleanup in the SNI layer itself. 
Had the disposed been removed in native stack, there should have been a lot of mem leaks.

Gary-Zh commented:
@Wraith2  Please look into `Microsoft.EntityFrameworkCore.SequenceEndToEndTest.Can_use_sequence_end_to_end_from_multiple_contexts_concurrently_async` since this one is always reproducible. I ran managed SNIPacket on Windows. Do you have EFCore test local setup? If so then you can debug into the test case.

Wraith2 commented:
I have the repo to run the tests but I've never used EF and that test case looks complicated. 
As I've said a couple of times now I have seen this problem before directly in the manual tests, fully reproducible. I couldn't resolve it and couldn't isolate the problem so I backed away from the cause to the current code changes because they don't exhibit the problem. It seems to be the case that the error is in something fairly fundamental that I've changed and I don't know where that is.

I know I can't fix this alone because I've tried. Trying to trace individual packets through this mountain of a library is virtually impossible. If it were my job then i've give it a go but this isn't my job. If there's not going to be any attempt at collaboration or suggestion on avenues of approach from the expert owners then there's no point leaving this open.

saurabh500 commented:
Hey @Wraith2 don't close this PR. Let's add the do not merge tag to it and revisit it when there is bandwidth. I agree that it is not easy to debug through the mountain of packets. 

Please reopen this PR.

saurabh500 commented:
Also please note that at this point, we will also have to debug into this problem and figure out where things are going wrong. 
The packet handling is quite complex and we have some idea of what is going on, but not to the last bit of nuances involved. You making the effort to improve it is absolutely appreciated and this PR will take time to push through and figure out the problems with what is going wrong in the improvments that you are suggesting.

Gary-Zh commented:
Hey @Wraith2 , sorry to make you depressed about what happened yesterday. We don't want to break your head debugging, trying to resolve the existing issue in corefx that is not related to this PR. It was just according to the test result, we thought the issue was brought by your change. 
Good news is that it looks the timeout issue is not from your change. We did regression test by building .dll file with your changes, but we compared the test result to an existing System.Data.SqlClient.dll which actually uses unmanaged SNIPacket.
After testing with managed code, that failure is still reproducible without your changes. We'll spend more time investigating the issue also move on to review and merge this PR.

saurabh500 commented on src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNIPacket.cs:
```diff
@@ -90,68 +78,54 @@ public void InvokeCompletionCallback(uint sniErrorCode)
         /// </summary>
         /// <param name="capacity">Length of byte array to be allocated</param>
-        public void Allocate(int capacity)
+        public void Allocate(int capacity, bool reserveMuxHeader)
```
Allocate is called from constructor only. This method can be made private.

saurabh500 commented on src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNIPacket.cs:
```diff
@@ -90,68 +78,54 @@ public void InvokeCompletionCallback(uint sniErrorCode)
+            SNIPacketFlags flags = reserveMuxHeader ? SNIPacketFlags.MuxHeaderReserved : SNIPacketFlags.None;
+            int headerCapacity = reserveMuxHeader ? SNISMUXHeader.HEADER_LENGTH : 0;
+            int totalCapacity = headerCapacity + capacity;
+            if (_data != null)
```
_data will always be null at this point, since Allocate is called from the constr. This path of checking if _Data is null or not, is not needed.

Wraith2 replied:
Simplified the entire file. Allocate could be moved inline in the ctor, worth doing?

saurabh500 commented on src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNIPacket.cs:
```diff
@@ -90,68 +78,54 @@ public void InvokeCompletionCallback(uint sniErrorCode)
-                _data = null;
             }
 
             if (_data == null)
```
_data is null at this point, you can get rid of the check,.

Wraith2 replied:
Done.

saurabh500 commented on src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNIPacket.cs:
```diff
@@ -90,68 +78,54 @@ public void InvokeCompletionCallback(uint sniErrorCode)
-                _data = ArrayPool<byte>.Shared.Rent(capacity);
-                _isBufferFromArrayPool = true;
+                _data = ArrayPool<byte>.Shared.Rent(totalCapacity);
+                flags |= SNIPacketFlags.ArrayFromPool; // set local not instance because it will be assigned after this block
```
I don't see _data being allocated using a new. It will always be from the array pool. Any reason why we need this flag to signify that the SNIPacket is from the ArrayPool?

Wraith2 replied:
Flags area all removed now. 

 have considered having a specific arraypool for packet buffers. The lifetime of the buffers we use isn't quite as temporary as other allocations. We know there will only be a small number of packets (roughly 2-3, in, out and attn) per connection active at any one time and in the common case of a single connection string packet size per application that would means the pool is small and only contains arrays of a specific size.

saurabh500 commented on src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNIPacket.cs:
```diff
@@ -12,37 +13,28 @@ namespace System.Data.SqlClient.SNI
+        private int _length; // the length of the data in the data segment, advanced by Append-ing data, does not include smux header length
+        private int _capacity; // the total capacity requested, if the array is rented this may be less than the _data.Length, does not include smux header length
+        private int _offset; // the start point of the data in the data segment, advanced by Take-ing data
+        private int _header; // the amount of space at the start of the array reserved for the smux header, this is zeroed in SetHeader
```
THe name `_header` sounds more like data than a number. Can you rename this to `_headerSize`

Wraith2 replied:
Renamed throughout. headerLength, dataLength, dataOffset, dataCapacity. It reads better to me, see if you agree or have other naming suggestions.

saurabh500 commented on src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNIPacket.cs:
```diff
@@ -211,36 +175,51 @@ public int TakeData(byte[] buffer, int dataOffset, int size)
+            header.Write(_data.AsSpan(0, SNISMUXHeader.HEADER_LENGTH));
+            _capacity += _header;
+            _length += _header;
+            _header = 0;
```
What does `_header = 0 ` mean ? Does it mean that there is no more header left to be read? If yes, then we need to come up with a better name for this var. Something that shows that it contains the length of unread header.

Wraith2 replied:
changed to headerLength.

saurabh500 commented on src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNIPacket.cs:
```diff
@@ -211,36 +175,51 @@ public int TakeData(byte[] buffer, int dataOffset, int size)
+            Debug.Assert(((_flags & SNIPacketFlags.MuxHeaderReserved) == SNIPacketFlags.MuxHeaderReserved), "writing mux heaser to non-mux packet");
+
+            header.Write(_data.AsSpan(0, SNISMUXHeader.HEADER_LENGTH));
+            _capacity += _header;
```
Why is the _capacity being incremented by the size of the _header? Sounds like you are incrementing this to set the new offset of capacity? 
Basically the name of the member int and what it is doing seem out of sync

saurabh500 replied:
It seems like when the SMUX header is set, you are telling this packet that now your capacity is original capacity + the header, that is written by MUX and hence the length will increase as well.

Wraith2 replied:
Once you allocate the header it's part of the data that needs to be sent so it's length is added to the total data. I chose the model where the header is a reserved space but once used it's part of the packet, it made thinking about the muxheader as a part of the packet easier. The renaming has made this feel clearer to me, see what you think.

saurabh500 commented on src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNIPacket.cs:
```diff
@@ -58,15 +50,11 @@ public string Description
-        }
+        public bool MuxHeaderReserved => ((_flags & SNIPacketFlags.MuxHeaderReserved) == SNIPacketFlags.MuxHeaderReserved);
+
+        public bool MuxHeaderWritten => ((_flags & SNIPacketFlags.MuxHeaderWritten) == SNIPacketFlags.MuxHeaderWritten);
```
I dont see this Property being read anywhere. Why is this needed?

Wraith2 replied:
Both removed. There's a new ReservedHeaderSize field now.

saurabh500 commented on src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNIPacket.cs:
```diff
@@ -211,36 +175,51 @@ public int TakeData(byte[] buffer, int dataOffset, int size)
+
+            Debug.Assert(_offset == 0, "writing mux header to partially read packet");
+            Debug.Assert(_header == SNISMUXHeader.HEADER_LENGTH, "writing mux header to partially incorrectly sized reserved region");
+            Debug.Assert(((_flags & SNIPacketFlags.MuxHeaderReserved) == SNIPacketFlags.MuxHeaderReserved), "writing mux heaser to non-mux packet");
```
Nit: typo heaser -> header

Wraith2 replied:
Fixed.

saurabh500 commented on src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNIPacket.cs:
```diff
@@ -90,68 +78,54 @@ public void InvokeCompletionCallback(uint sniErrorCode)
             }
 
+            _flags = flags;
             _capacity = capacity;
```
It makes more sense to have _capacity = totalCapacity instead of capacity. 
In case the totalCapacity has the header length in it, it is certain that MARS Handle will write the header before writing out the packet that it needs to send. And this way you wont have to change the capacity in the `SetHeader` method below.

saurabh500 commented on src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNIPacket.cs:
```diff
@@ -211,36 +175,51 @@ public int TakeData(byte[] buffer, int dataOffset, int size)
+        /// Set the MARS SMUX header information for this packet
+        /// </summary>
+        /// <param name="header">the header to write into the packet</param>
+        public void SetHeader(SNISMUXHeader header)
```
Ideally I would like to have SNIPacket agnostic of MARS. It should be a container for the bytes and have the information about the stream that it should write to. 
Can you refactor this, so that SNIPacket is agnostic?

Wraith2 replied:
I've split this into two functions `GetHeadrBuffer` and `SetHeaderActive` because the only way to do it with a single function i could see uses a delegate and i don't want to pay the allocation cost.  It's used from the mars handle now.

To remove the knowledge of smux from other places i've changed the SNIHandle so that it has a reserved header property which indicates how large a header it wants, all standard handles return 0 and the marsh handle returns the smux header length so the handle now determines the header reservation and no purpose is specified. These changes flow through the other files.

saurabh500 replied:
Thanks, this is much nicer approach.

saurabh500 commented on src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNIPacket.cs:
```diff
@@ -183,21 +157,11 @@ public void AppendData(ReadOnlySpan<byte> data)
         }
```
This method `public void AppendData(ReadOnlySpan<byte> data)`  is unused now. This can be deleted.

Wraith2 replied:
Done.

Wraith2 committed:
Merge remote-tracking branch 'dotnet/master' into sqlperf-managedsmux

Wraith2 committed:
address feedback

Wraith2 commented:
Feedback addressed. function and manual tests pass.

saurabh500 commented on src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNIPacket.cs:
```diff
@@ -3,70 +3,44 @@
 // See the LICENSE file in the project root for more information.
 
 using System.Buffers;
```
Much cleaner now !! :) Thank you

saurabh500 reviewed (approved):
LGTM. Pending the CI for Linux not showing any regressions.
@Gary-Zh  is going to work on that.

saurabh500 commented:
@Wraith2 This comment is not related to this PR.

The test above `Microsoft.EntityFrameworkCore.SequenceEndToEndTest.Can_use_sequence_end_to_end_from_multiple_contexts_concurrently_async` has been failing quite frequently. 
It has boiled down to the SqlClient not being able to spin up connections from the pool quickly enough, with the managed SNI.

The bug in EF is https://github.com/aspnet/EntityFrameworkCore/issues/5254

There was a simpler repro filed for SqlClient at https://github.com/dotnet/corefx/issues/8701 which I had looked at and couldn't repro when I had closed the bug. 

This might be another area to look at to see what is preventing SQLClient connection pool connection retrieval to be slow. 

Just putting it out there as a known issue for SqlClient which can use some help. 

I know that you had profiled the connection pool and made changes where the connection is retrieved using Integrated Authentication.

karelz commented:
@Gary-Zh do you have ETA for the test verification? It's been 9 days since that info ...

saurabh500 commented:
There is an issue with EF tests hanging on Linux which is being followed up on https://github.com/aspnet/EntityFrameworkCore/issues/15333
We use 
Also @Wraith2 found out that there is an additional test from DataAccessBenchmark that helped catch issues which were missed by the tests that we run. There was a bug that went through, which was caught by DataAccessBenchmark when @Wraith2 executed those tests. Once we port those over and get some help from the EF team about the test hang issue, we can move forward.

karelz commented:
Thanks for update! Do you have rough ETA? (just to set overall expectations)

Wraith2 commented:
I've got a further branch of this which adds packet caching per connection and reliably fails a manual test with a timeout, probably the same one you're seeing in the EF core tests. I meant to add this functionality originally but the presence of the timeout every time I added the cache convinced me I was doing something wrong but couldn't find it. 

Now we know that it isn't my change causing it you might find it useful to use this branch to try and track it down, as I said I tried for some weeks and couldn't. https://github.com/Wraith2/corefx/tree/sqlperf-managedsmux2

Wraith2 commented:
In my local copy of the branch I linked I seem to have fixed the timeout. It wasn't my intention to do so but another bug where a packet was used after it was no longer valid became apparent when packet recycling was introduced. 

The error I fixed was in `SNITcpHandle.Receive`, it returns a packet as an out parameter but if an error occurs that out parameter is not set to null. Changing that method so that errors always return null removed the error I was looking into (which was in a similar test) and the timeout I'd seen before. Might be worth trying.

ViktorHofer commented:
@saurabh500 @Wraith2 could you please provide a status on this?

karelz commented:
@AfsanehR @Gary-Zh what is status of this PR? Any update? Was there some offline deal with @Wraith2 about this one?

Gary-Zh commented:
Hi @karelz , most of the tests have already passed, there is only one pending test: EFCore on Linux, and it keeps hanging on all of our Linux VMs.
As suggested by the EFCore team we've tried to increase the number of logic CPUs allocated to the VM but it does not change the result. Once this issue is resolved and test passes, we'll merge the PR immediately.

Here's the link to the hanging issue:
https://github.com/aspnet/EntityFrameworkCore/issues/15333

karelz commented:
Thanks @Gary-Zh for the update. Who is working on the last failing test? You / SqlClient team / @Wraith2?

Gary-Zh commented:
Hi @karelz , no test failures spotted yet because we can't even run the tests successfully. The EFCore test itself is hanging and we are looking for help from EFCore team right now.
It passed on windows tho, with & without native SNI, so I think it should pass on Linux as well.
However we still want to see it passing on Linux before merging this PR.

karelz commented:
Thanks @Gary-Zh! If you have rough ETA, or at least time-frame in which you plan to reply back, that would be useful to @Wraith2 and me as well :)

cheenamalhotra commented:
**Note:** Porting to Microsoft.Data.SqlClient incomplete without #40732
